drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1114 @@
+
+	/* draw on all outputs */
+	for (iter = output_list; iter; iter = iter->next) {
+		iter->r = rand() % 0xff;
+		iter->g = rand() % 0xff;
+		iter->b = rand() % 0xff;
+		iter->r_up = iter->g_up = iter->b_up = true;
+
+		/* here we have the atomic commit that leads to page-flips */
+		modeset_draw_out(fd, iter);
+	}
+
+	/* wait 5s for VBLANK or input events */
+	while (time(&cur) < start + 5) {
+		FD_SET(0, &fds);
+		FD_SET(fd, &fds);
+		v.tv_sec = start + 5 - cur;
+
+		ret = select(fd + 1, &fds, NULL, NULL, &v);
+		if (ret < 0) {
+			fprintf(stderr, "select() failed with %d: %m\n", errno);
+			break;
+		} else if (FD_ISSET(0, &fds)) {
+			fprintf(stderr, "exit due to user-input\n");
+			break;
+		} else if (FD_ISSET(fd, &fds)) {
+			drmHandleEvent(fd, &ev);
+		}
+	}
+}
+
+/*
+ * modeset_cleanup() stays the same.
+ */
+
+static void modeset_cleanup(int fd)
+{
+	struct modeset_out *iter;
+	drmEventContext ev;
+	int ret;
+
+	/* init variables */
+	memset(&ev, 0, sizeof(ev));
+	ev.version = DRM_EVENT_CONTEXT_VERSION;
+	ev.page_flip_handler = modeset_page_flip_event;
+
+	while (output_list) {
+		/* remove from global list */
+		iter = output_list;
+		output_list = iter->next;
+
+		/* if a page-flip is pending, wait for it to complete */
+		iter->cleanup = true;
+		fprintf(stderr, "wait for pending page-flip to complete...\n");
+		while (iter->pflip_pending) {
+			ret = drmHandleEvent(fd, &ev);
+			if (ret)
+				break;
+		}
+
+		/* restore saved CRTC configuration */
+		if (!iter->pflip_pending)
+			drmModeSetCrtc(fd,
+				       iter->saved_crtc->crtc_id,
+				       iter->saved_crtc->buffer_id,
+				       iter->saved_crtc->x,
+				       iter->saved_crtc->y,
+				       &iter->conn_id,
+				       1,
+				       &iter->saved_crtc->mode);
+		drmModeFreeCrtc(iter->saved_crtc);
+
+		/* destroy framebuffers */
+		modeset_destroy_fb(fd, &iter->bufs[1]);
+		modeset_destroy_fb(fd, &iter->bufs[0]);
+
+		/* free allocated memory */
+		free(iter);
+	}
+}
+
+/*
+ * main() also changes. Instead of performing the KMS setup calling
+ * drmModeSetCrtc(), we instead setup it using the atomic API with
+ * the function modeset_perform_modeset().
+ */
+
+int main(int argc, char **argv)
+{
+	int ret, fd;
+	const char *card;
+	struct modeset_out *iter;
+	drmModeAtomicReq *req;
+
+	/* check which DRM device to open */
+	if (argc > 1)
+		card = argv[1];
+	else
+		card = "/out/dri/card0";
+
+	fprintf(stderr, "using card '%s'\n", card);
+
+	/* open the DRM device */
+	ret = modeset_open(&fd, card);
+	if (ret)
+		goto out_return;
+
+	/* prepare all connectors and CRTCs */
+	ret = modeset_prepare(fd);
+	if (ret)
+		goto out_close;
+
+	/* save crtc object, so we can restore it later */
+	for (iter = output_list; iter; iter = iter->next)
+		iter->saved_crtc = drmModeGetCrtc(fd, iter->crtc_id);
+
+	/* perform modeset using atomic commit */
+	req = drmModeAtomicAlloc();
+	modeset_perform_modeset(fd, req);
+	drmModeAtomicFree(req);
```

Comment:
Because you put the loop over all outputs inside `modeset_perform_modeset()`, now there is no reason to have the atomic request outside here, it should be inside as well.

Reply:
Done.